cmd/options.go

```diff
@@ -235,18 +235,19 @@ func getOptions(flags *pflag.FlagSet) (lib.Options, error) {
 		return opts, err
 	}
 
-	if len(runTags) > 0 {
-		parsedRunTags := make(map[string]string, len(runTags))
-		for _, s := range runTags {
-			var name, value string
-			name, value, err = parseTagNameValue(s)
-			if err != nil {
-				return opts, fmt.Errorf("error parsing tag '%s': %w", s, err)
-			}
-			parsedRunTags[name] = value
+	// Atlas initialization where the root node is created.
+	// All the new tagsets must branch out from the root,
+	// so the underhood structure can minimize the memory footprint allocating the nodes only once.
+	tagsRootNode := metrics.RootTagMap()
```

Comment:
I think having this dynamic atlas-based thing in `lib.Options` might be a mistake, or at least an over-complication we don't need :thinking: We only need the option validation and string values here, a `map[string]string` (or some sorted wrapper around one) should be plenty good enough for the Options, we just need to validate the values at this point... The atlas integration and its root node seems like it should be stored somewhere outside of the Options, in some sort of a test state? :thinking: 

So, on Thursday, when I started reviewing this, I though this place was the already existing `lib.RuntimeState` that was initialized in `cmd/`, `loadedTest.initializeFirstRunner()`, where we also initialize the `metrics.Registry` and the built-in metrics: 
https://github.com/grafana/k6/blob/f2684ad90fb7cfc8e02cfa776d303e3c9df10f83/lib/runtime_state.go#L10-L18
https://github.com/grafana/k6/blob/f2684ad90fb7cfc8e02cfa776d303e3c9df10f83/cmd/test_load.go#L112-L117

That gets propagated in the `js.Runner` (though it probably needs a bit of refactoring to actually save the whole object instead of deconstruct it): https://github.com/grafana/k6/blob/f2684ad90fb7cfc8e02cfa776d303e3c9df10f83/js/runner.go#L109

Unfortunately, this gets initialized before we have the consolidated and derived options, and we need that to get the final value of `--tags` :sob: https://github.com/grafana/k6/blob/f2684ad90fb7cfc8e02cfa776d303e3c9df10f83/cmd/test_load.go#L94

I originally thought "Oh, #883 strikes again :sob: ", but then reconsidered :sweat_smile: This isn't a matter of configuration, precisely, it's a matter of some state that we can't propagate to the rest of the test because of our poor code structure :disappointed: So... I kind of went down that rabbit hole and fixed the problem, and will submit a refactoring PR shortly (just fixing some final tests).

Reply:
>We just need to validate and store the values at this point, not start building the root atlas Node... The atlas integration and its root Node seems like it should be stored somewhere outside of the Options, in some sort of a test state?

I already have this version in local with a different approach, sorry if I haven't yet pushed it but the last transition (`output/cloud/...`) takes time.

Instead of using the `RuntimeState`, I added a method on the Registry for getting the root TagMap. I did it in this way so it wasn't required to fix more tests initializing a new dependency. For example, [in this method ](https://github.com/grafana/k6/blob/f2684ad90fb7cfc8e02cfa776d303e3c9df10f83/metrics/engine/engine.go#L89-L92) we need the root for initializing a new `SampleTags` for the Submetric and the Registry is already available in the metrics engine.
We can consider having a dedicated field in the RuntimeState if we prefer.

Reply:
hmm, why do we need the root atlas Node to create a sub-metric? :confused:

Reply:
We have a [SampleTags initialization](https://github.com/grafana/k6/blob/f2684ad90fb7cfc8e02cfa776d303e3c9df10f83/metrics/metric.go#L102-L108) in AddSubmetric.

Reply:
Ah, I see, since you're using [`Node.Contains(Node)`](https://github.com/mstoykov/atlas/blob/23cc720eb639c314be5df4442d8567969c63fc06/atlas.go#L101) to see if a time series' tags contain the sub-metric tags, they need to have a common root. For that, I guess storing it in the `metrics.Registry` is great :+1:  :+1: 

But I'd suggest against storing the `Node` that results from applying the `--tags` there. Let me publish my refactoring PR today and you'll see why I think there is a better place.

Reply:
So, this is my proposal: https://github.com/grafana/k6/pull/2627 _(most of the diff is just fixes in the tests, the actual code changes are fairly minor)_

While the base `atlas.New()` root `Node` can sit in the `metrics.Registry` (though it can also be in the `TestPreInitState` :man_shrugging: ), the Node that is the result of `--tags` can be saved in the `TestRunState` object. The executors will then be able to build on top of that and add the `scenario` tag, as well as any `tags` that were explicitly specified in the scenario config. I think all of this won't need any `SampleTags` or locking, just simple atlas nodes in `TestRunState` and the scenarios, and simple string:string map for `lib.Options` (or a simple sorted wrapper without any locking).

Reply:
While I agree with @na-- on how the atlas node/tags should be handled and kept (I personally only thing you need the non read-only AddTag in places where you are emitting metrics) I don't see why we need to refactor such a huge part of k6 to do it. Especially given that it's the summer and a lot of us will be on/off quite a bit and there are at least 3 big changes in the work which will definitely need to be rebased on that quite big PR ...

Reply:
I didn't intend the PR to become so big in terms of LoC, but that is just because of the tests... The actual changes in the non-test code are fairly few and on the small side, basically no logical changes anywhere, just moving things around.